File: packages/sentinel-js/src/trace.ts
import { validateSchema } from './schema'

export interface TracePayload {
  traceId: string
  spanId: string
  parentSpanId?: string
  name: string
  startTime: string
  endTime: string
  attributes?: Record<string, unknown>
}

export interface ToolCallResult {
  valid: boolean
  errors: string[]
}

async function emitSpan(payload: TracePayload, endpoint: string, apiKey: string): Promise<void> {
  try {
    const url = `${endpoint.replace(/\/+$/, '')}/api/traces`
    await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(payload),
    })
  } catch {
    // never let a dead or malformed Sentinel endpoint crash the caller's agent
  }
}

export class Trace {
  readonly traceId: string
  readonly spanId: string
  private readonly name: string
  private readonly startTime: string
  private readonly endpoint: string
  private readonly apiKey: string
  private readonly project: string

  constructor(name: string, endpoint: string, apiKey: string, project: string) {
    this.name = name
    this.endpoint = endpoint
    this.apiKey = apiKey
    this.project = project
    this.traceId = crypto.randomUUID().replace(/-/g, '')
    this.spanId = crypto.randomUUID().replace(/-/g, '')
    this.startTime = new Date().toISOString()
  }

  async end(attributes: Record<string, unknown> = {}): Promise<void> {
    const endTime = new Date().toISOString()
    await emitSpan(
      {
        traceId: this.traceId,
        spanId: this.spanId,
        name: this.name,
        startTime: this.startTime,
        endTime,
        attributes: { project: this.project, ...attributes },
      },
      this.endpoint,
      this.apiKey
    )
  }

  async toolCall(
    name: string,
    declaredSchema: Record<string, unknown>,
    actualParameters: Record<string, unknown>
  ): Promise<ToolCallResult> {
    const errors = validateSchema(declaredSchema, actualParameters)
    const result: ToolCallResult = { valid: errors.length === 0, errors }
    const now = new Date().toISOString()
    await emitSpan(
      {
        traceId: this.traceId,
        spanId: crypto.randomUUID().replace(/-/g, ''),
        parentSpanId: this.spanId,
        name: `tool_call:${name}`,
        startTime: now,
        endTime: now,
        attributes: {
          project: this.project,
          toolName: name,
          declaredSchema,
          actualParameters,
          valid: result.valid,
          errors: result.errors,
        },
      },
      this.endpoint,
      this.apiKey
    )
    return result
  }
}

File: packages/sentinel-py/sentinel/_trace.py
import json
import uuid
from datetime import datetime, timezone
from typing import Optional
from urllib.error import URLError
from urllib.request import Request, urlopen

from ._client import SentinelConfig, get_config


def _now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def _emit_span(payload: dict, config: SentinelConfig) -> None:
    try:
        url = f"{config.endpoint.rstrip('/')}/api/traces"
        request = Request(
            url,
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {config.api_key}",
            },
            method="POST",
        )
        urlopen(request, timeout=5)
    except (URLError, ValueError):
        pass


class Trace:
    def __init__(self, name: str):
        self._config = get_config()
        self.name = name
        self.trace_id = uuid.uuid4().hex
        self.span_id = uuid.uuid4().hex
        self._start_time: Optional[str] = None

    def __enter__(self) -> "Trace":
        self._start_time = _now_iso()
        return self

    def __exit__(self, exc_type, exc_value, traceback) -> bool:
        end_time = _now_iso()
        _emit_span(
            {
                "traceId": self.trace_id,
                "spanId": self.span_id,
                "name": self.name,
                "startTime": self._start_time,
                "endTime": end_time,
                "attributes": {"project": self._config.project},
            },
            self._config,
        )
        return False


def trace(name: str) -> Trace:
    return Trace(name)
